x/bsonx/bsoncore/document.go

```diff
@@ -291,6 +294,89 @@ func (d Document) String() string {
 	return buf.String()
 }
 
+// StringN stringifies a document upto N bytes
+func (d Document) StringN(n int) string {
+	if len(d) < 5 {
+		return ""
+	}
+	var buf strings.Builder
+	buf.Grow(n) //Preallocate buffer size
+	buf.WriteByte('{')
+
+	length, rem, _ := ReadLength(d)
+	length -= 4
+
+	var elem Element
+	var ok bool
+
+	first := true
+
+	n -= buf.Len() + 1
+	if n > 0 {
+		for length > 1 {
+			if !first {
+				buf.WriteByte(',')
+			}
+			elem, rem, ok = ReadElement(rem)
+			length -= int32(len(elem))
+			if !ok {
+				return ""
+			}
+
+			str := elem.StringN(n)
+			if buf.Len()+len(str) > n {
+				truncatedStr := truncate(str, uint(n-buf.Len()))
+				buf.WriteString(truncatedStr)
+				break
+			}
+
+			buf.WriteString(str)
+			first = false
+		}
+	}
+
+	if buf.Len()+1 <= n {
+		buf.WriteByte('}')
+	} else {
+		buf.WriteString(TruncationSuffix)
+	}
+
+	return buf.String()
+}
+
+// truncate truncates a given string for a certain width
+func truncate(str string, width uint) string {
+	if width == 0 {
+		return ""
+	}
+
+	if len(str) <= int(width) {
+		return str
+	}
+
+	// Truncate the byte slice of the string to the given width.
+	newStr := str[:width]
+
+	// Check if the last byte is at the beginning of a multi-byte character.
+	// If it is, then remove the last byte.
+	if newStr[len(newStr)-1]&0xC0 == 0xC0 {
+		return newStr[:len(newStr)-1] + TruncationSuffix
+	}
+
+	// Check if the last byte is in the middle of a multi-byte character. If
+	// it is, then step back until we find the beginning of the character.
+	for i := len(newStr) - 1; i >= 0; i-- {
+		if (newStr[i] & 0xC0) != 0x80 {
```

Comment:
This line tests if the string _is not_ in the middle of a multi-byte character. See [this](https://stackoverflow.com/a/3911566) explanation. In this case, you don't need to do anything. Suggest using the following block instead of this for loop:

```go
// Check if the last byte is a multi-byte character
if newStr[len(newStr)-1]&0xC0 == 0x80 { 
	// If it is, step back until you we are at the start of a character
	for i := len(newStr) - 1; i >= 0; i-- { 
		if newStr[i]&0xC0 == 0xC0 {
			// Truncate at the end of the character before the character we stepped back to
			return newStr[:i]
		}
	}
}
```

Taking an example from the[ UTF-8 wiki](https://en.wikipedia.org/wiki/UTF-8#Examples), `𨉟呐㗂越` is 13 bytes. The last two MBCs `㗂` and `越` are each 3 bytes. So the above algorithm with n=9 should result in `𨉟呐`. The steps are as follows:

1. newStr is calculated by truncating to 9 bytes: `𨉟呐��` where ?? are the first two bytes of `㗂`
2. The last byte is `151` which is not the beginning of a multi byte character (1110, from the encoding section of the wiki)
3. So we move to the next byte `227` which is the beginning of a multi byte character.
4. We remove 227 and 151, leaving `𨉟呐`.

Reply:
Thank you, this makes a lot more sense!